edgekv/src/datastore.rs

```
@@ -0,0 +1,835 @@
+            .keys
+            .write()
+            .map_err(|e| EdgeKVError::RWLockPoisonError(format!("{}", e)))?;
+        keys_dir_writer.clear();
+        Ok(())
+    }
+
+
+    //TODO change to use references
+    pub fn keys(&self) -> Vec<Vec<u8>> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return vec![];
+            }
+        };
+
+        keys_dir_reader.iter().map(|(k, _)| k.clone()).collect()
+    }
+
+    //TODO change to use references
+    pub fn range<R>(&self, range : R) -> Vec<Vec<u8>> where R : RangeBounds<Vec<u8>> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return vec![];
+            }
+        };
+        keys_dir_reader.range(range).map(|(k, _)| k.clone()).collect()
+    }
+
+    //TODO change to use references
+    pub fn prefix(&self, prefix: &Vec<u8>) -> Vec<Vec<u8>> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return vec![];
+            }
+        };
+        keys_dir_reader
+            .range(prefix.clone()..)
+            .take_while(|(k, _)| k.starts_with(prefix))
+            .map(|(k, _)| k.clone())
+            .collect()
+    }
+
+
+    pub fn get(&self, key: &[u8]) -> Option<KeyDirEntry> {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
+                return None;
+            }
+        };
+        match keys_dir_reader.get(key) {
+            None => None,
+            Some(entry) => {
+                if let Persisted(entry) = entry {
+                    return Some(entry.clone());
+                }
+                return None;
+            }
+        }
+    }
+
+    pub fn size(&self) -> usize {
+        let keys_dir_reader = match self.keys.read() {
+            Ok(rdr) => rdr,
+            Err(_) => {
```

Same here, should these be silently ignored?